libraries/AC_AttitudeControl/AC_AttitudeControl_Heli.cpp

```diff
@@ -598,4 +598,64 @@ void AC_AttitudeControl_Heli::set_inverted_flight(bool inverted)
+{
+    // convert thrust vector to a roll and pitch angles
+    // this negates the advantage of using thrust vector control, but works just fine
+    Vector3f angle_target = attitude_from_thrust_vector(thrust_vector, _ahrs.yaw).to_vector312();
+
+    float euler_roll_angle_cd = degrees(angle_target.x) * 100.0f;
+    if (_inverted_flight) {
+        euler_roll_angle_cd = wrap_180_cd(euler_roll_angle_cd + 18000);
+        AC_AttitudeControl::input_euler_angle_roll_pitch_euler_rate_yaw(euler_roll_angle_cd, degrees(angle_target.y) * 100.0f, heading_rate_cds);
+    } else {
+        AC_AttitudeControl::input_thrust_vector_rate_heading(thrust_vector, heading_rate_cds, slew_yaw);
+    }
+
 }
+
+// Command a thrust vector, heading and heading rate
+void AC_AttitudeControl_Heli::input_thrust_vector_heading(const Vector3f& thrust_vector, float heading_angle_cd, float heading_rate_cds)
+{
+    // convert thrust vector to a roll and pitch angles
+    Vector3f angle_target = attitude_from_thrust_vector(thrust_vector, _ahrs.yaw).to_vector312();
+
+    float euler_roll_angle_cd = degrees(angle_target.x) * 100.0f;
+    if (_inverted_flight) {
+        euler_roll_angle_cd = wrap_180_cd(euler_roll_angle_cd + 18000);
+        // note that we are throwing away heading rate here
+        AC_AttitudeControl::input_euler_angle_roll_pitch_yaw(euler_roll_angle_cd, degrees(angle_target.y) * 100.0f, heading_angle_cd, true);
+    } else {
+        AC_AttitudeControl::input_thrust_vector_heading(thrust_vector, heading_angle_cd, heading_rate_cds);
+    }
+
+}
+
+void AC_AttitudeControl_Heli::input_thrust_vector_heading(const Vector3f& thrust_vector, HeadingCommand heading)
+{
+    // convert thrust vector to a roll and pitch angles
+    // this negates the advantage of using thrust vector control, but works just fine
+    Vector3f angle_target = attitude_from_thrust_vector(thrust_vector, _ahrs.yaw).to_vector312();
+
+    float euler_roll_angle_cd = degrees(angle_target.x) * 100.0f;
+    if (_inverted_flight) {
+        euler_roll_angle_cd = wrap_180_cd(euler_roll_angle_cd + 18000);
+        switch (heading.heading_mode) {
+        case HeadingMode::Rate_Only:
+            AC_AttitudeControl::input_euler_angle_roll_pitch_euler_rate_yaw(euler_roll_angle_cd, degrees(angle_target.y) * 100.0f, heading.yaw_rate_cds);
+            break;
+        case HeadingMode::Angle_Only:
+            AC_AttitudeControl::input_euler_angle_roll_pitch_yaw(euler_roll_angle_cd, degrees(angle_target.y) * 100.0f, heading.yaw_angle_cd, true);
+            break;
+        case HeadingMode::Angle_And_Rate:
+            AC_AttitudeControl::input_euler_angle_roll_pitch_yaw(euler_roll_angle_cd, degrees(angle_target.y) * 100.0f, heading.yaw_angle_cd, true);
+            break;
+        }
+    } else {
+        AC_AttitudeControl::input_thrust_vector_heading(thrust_vector, heading);
+    }
+}
```

Comment:
I don't think you need this. The base method call through the `input_thrust_vector_rate_heading` and `input_thrust_vector_heading` which are overridden already.

Reply:
deleted. thanks.